### ISSUE TITLE
[Phase 4] Add INT8 Quantization Support for 4x Memory Reduction

### ISSUE THREAD

**Issue by ccchow** [ccchow/zo2#2]
## 🎯 Objective
Implement INT8 quantization to achieve 4x memory reduction for model weights while maintaining training quality.

## 📊 Context
Current memory usage for OPT models shows opportunity for quantization:
- OPT-2.7B uses 3,670 MB GPU memory
- With INT8: Target <1,000 MB GPU memory
- Enables training of much larger models (OPT-6.7B+)

## 📝 Implementation Tasks
- [ ] Add quantization utilities in `zo2/utils/quantization.py`
- [ ] Implement INT8 weight quantization/dequantization
- [ ] Integrate with MeZO gradient estimation
- [ ] Add `quantization_config` to `ZOConfig`
- [ ] Implement mixed-precision ZO updates
- [ ] Create quantization-aware training logic

## 🏗️ Architecture Changes
```python
class QuantizationConfig:
    enabled: bool = False
    dtype: str = "int8"  # int8, int4 future support
    quantize_embeddings: bool = False
    quantize_lm_head: bool = False
    calibration_samples: int = 1000
```

## 🎯 Success Criteria
- [ ] Memory reduction ≥3.5x for weights
- [ ] Accuracy retention ≥95% on downstream tasks
- [ ] Training convergence maintained
- [ ] Support for OPT-6.7B on 24GB GPU

## 🧪 Test Plan
1. **Memory Benchmarks**
   - Measure memory before/after quantization
   - Track memory during training

2. **Accuracy Validation**
   - Fine-tune on SST-2 dataset
   - Compare with FP16 baseline
   - Measure perplexity degradation

3. **Scaling Tests**
   - OPT-2.7B (baseline)
   - OPT-6.7B (newly enabled)
   - OPT-13B (stretch goal)

## 📈 Expected Impact
- 4x memory reduction for weights
- Enable 6.7B+ models on consumer GPUs
- Maintain 90-95% of FP16 accuracy

## 🔗 References
- Research: "The case for 4-bit precision" (https://arxiv.org/abs/2212.09720)
- PyTorch Quantization: https://pytorch.org/docs/stable/quantization.html

**Priority**: High
**Timeline**: 1 week
